File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/PlayQueue.cs
namespace IteratorDesignPattern.Example3;

public class PlayQueue : IAggregate<Song>
{
    private readonly Queue<Song> _queue = new();

    public void Enqueue(Song song)
    {
        _queue.Enqueue(song);
    }

    public Queue<Song> GetQueue() => _queue;

    public IIterator<Song> CreateIterator()
    {
        return new QueueIterator(this);
    }
}

File: java/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/src/main/java/behavioral/iterator/example3/PlayQueue.java
package behavioral.iterator.example3;

import java.util.*;

public class PlayQueue implements IAggregate<Song> {
    private final Queue<Song> queue = new ArrayDeque<>();

    public void enqueue(Song song) { queue.add(song); }
    public Queue<Song> getQueue() { return queue; }

    @Override
    public IIterator<Song> createIterator() { return new QueueIterator(this); }
}
